app/controllers/courses_controller.rb

```diff
@@ -0,0 +1,250 @@
+    # Authorization:
+    # - If the course is still a draft then only available to course creator
+    # - If the course is not a draft then only available to moderator and admin
+    # Unauthorized access: Should get redirected to root (/)
+    # Authorized access: Replace the current course with that ID with the one submitted in the form. Return a message upon completion. Special note: only moderators and admins can set visibility to published.
+    # Return a JSON containing the status result of creation and the latest version of the course
+    def update
+        c_user = current_user()
+        return ( redirect_to :root ) unless logged_in?  # Ensure the user is logged in
+        course = Course.where(id: params[:course_id]).first()
+
+        # Course not found case
+        if course.nil?
+            return ( render :file => "#{Rails.root}/public/404.html", :status => 404 )
+        end
+
+        # Other Course case
+        if c_user.role == Role.admin or c_user.role == Role.moderator
+            status = course.update(course_params)
+            render :json => { "status": status, "course": course }
+
+        # Draft Course case
+        elsif course.visibility == Visibility.draft and course.user_id == c_user.id
+            if course_params[:visibility].to_i == Visibility.draft
+                status = course.update(course_params)
+                render :json => { "status": status, "course": course }
+            else
+                render :json => { "status": 400 }, :status => 400
+            end
+
+        # Not draft or authorized
+        else
+            redirect_to :root
+        end
+    end
+
+    # Request: DELETE /courses/(:course_id)
+    # Authorization:
+    # - If the course is still a draft then only available to course creator
+    # - If the course is not a draft then only available to moderator and admin
+    # Unauthorized access: Should get redirected to root (/)
+    # Authorized access: Remove the corresponding course from the database. Return message upon completion
+    def delete
+        c_user = current_user()
+        return ( redirect_to :root ) unless logged_in?  # Ensure the user is logged in
+        course = Course.where(id: params[:course_id]).first()
+
+        # Course not found case
+        if course.nil?
+            return ( render :file => "#{Rails.root}/public/404.html", :status => 404 )
+        end
+
+        # Draft Course case
+        if course.visibility == Visibility.draft and course.user_id == c_user.id
+            render :json => { "status": course.delete }
+        # Other Course case
+        elsif c_user.role == Role.admin or c_user.role == Role.moderator
+            render :json => { "status": course.delete }
+
+        # Not draft or authorized
+        else
+            redirect_to :root
+        end
+    end
+
+    private 
+        def course_params
+            params.require(:course).permit(:title, :user_id, :description, :tier, :visibility)
```

Comment:
user_id shouldn't be a user input. Courses are always created and owned by the session user, and we shouln't change the owner of a course.